### ISSUE TITLE
【LwM2M】【ATT certification】24 hour connection test failed

### ISSUE THREAD

**Issue by JameslaoA** [eclipse/leshan#863]
**_test case _[ATT] 23 24 Hour Connectivity_**L
The log is below:
2020.07.08 15:59:01.807 INFO  [DTLS] Client Key Exchange request received with DTLS identity: urn:imei:86537704000598
2020.07.08 15:59:01.923 ERROR [] Provisioning session error
Caused by: com.avsystem.ump.acs.provisioning.connector.lwm2m.ConnectionRejector$RejectionException: Transport binding's serialized value of dtlsSessionId: 5F05ED45486BF7609406B73DE6AF361A6B879755A55E67BF74EC9F7752548936 is different than expected 5F05E60FA3A579CAFEF4715BAA4A20B56FED3CA86AD3377A08E7D52EC330B792 

Check the AP log and find that the PSK has not changed.
Session ID changes (confirm that the network is normal).
Have you encountered this problem?How to solve this problem?

**Comment by sbernard31** [comment]
I'm sorry but I'm am lost here.
I have no context and you just provide me a log with an exception which is not raised by Leshan or even Californium : `com.avsystem.ump.acs.provisioning.connector.lwm2m.ConnectionRejector$RejectionException`

**Comment by JameslaoA** [comment]
I used your Leshan client to connect and interact with the server. However, during the 24-hour stable connection, the session ID changed and the test case failed

how to upload the log?

Thanks.

**Comment by sbernard31** [comment]
> how to upload the log?

https://docs.github.com/en/github/managing-your-work-on-github/file-attachments-on-issues-and-pull-requests

**Comment by sbernard31** [comment]
> I used your Leshan client 

You mean **leshan-client-demo** ?

> during the 24-hour stable connection, the session ID changed and the test case failed

If DTLS Connection is lost (or client think it may be lost), client could decide to re-establish a connection and so do a new full-handshake. In this case the sessionID will change.

I don't know why this could be a problem in your case.

**Comment by JameslaoA** [comment]
[SS808_24_HOUR_CONNECT_AP.txt](https://github.com/eclipse/leshan/files/4896024/SS808_24_HOUR_CONNECT_AP.txt)
[urn_imei_86537704000598_2020.07.07_2020.07.09.log](https://github.com/eclipse/leshan/files/4896026/urn_imei_86537704000598_2020.07.07_2020.07.09.log)

SS808_24_HOUR_CONNECT_AP.txt : leshan-client log 
urn_imei_86537704000598_2020.07.07_2020.07.09.log : server log

the issue time : 15:59:01.807 

please help to check it,
Thanks.

**Comment by JameslaoA** [comment]
![image](https://user-images.githubusercontent.com/17543107/87027864-5afcbe80-c210-11ea-8b93-e332e2a4e929.png)

If DTLS Connection is lost (or client think it may be lost), client could decide to re-establish a connection and so do a new full-handshake. In this case the sessionID will change.

I don't know why this could be a problem in your case.

--- 

please refer to the attachment picture, these is the ATT requirements.

the server log：

2020.07.08 15:59:02.267 ERROR [TASK, TEST_CASE] Finished executing test case "[ATT] 23 24 Hour Connectivity" with error. The device performed a new registration during the 24 hour observation period! This is prohibited in this test-case!
2020.07.08 15:59:02.267 INFO  [TASK] Finished executing task com.avsystem.ump.acs.task.xml.XmlTask with id: d:urn:imei:86537704000598:15 (test case "[ATT] 23 24 Hour Connectivity") with status: ERROR

Thanks.

**Comment by sbernard31** [comment]
> SS808_24_HOUR_CONNECT_AP.txt : leshan-client log

This doesn't looks at **leshan-client-demo** log at all.

> the issue time : 15:59:01.807

I see nothing at this time in "client log".

> The device performed a new registration during the 24 hour observation period! This is prohibited in this test-case!

Ok but doing a registration is normal for a LWM2M client ... 
This would be surprising but maybe the update request timeout (2min) and so client try to do a register, you could eventually increased the update timeout. 
See [DefaultRegistrationEngineFactory](https://github.com/eclipse/leshan/blob/leshan-1.0.2/leshan-client-core/src/main/java/org/eclipse/leshan/client/engine/DefaultRegistrationEngineFactory.java#L69)

**Comment by JameslaoA** [comment]
[pcap log.zip](https://github.com/eclipse/leshan/files/4896239/pcap.log.zip)
the pcap log.

**Comment by JameslaoA** [comment]
please the DefaultRegistrationEngine.java
the SS808_24_HOUR_CONNECT_AP.txt client  log：
07-08 11:59:01.259  2461  3534 D lihy lwm2m: Registration update failed: code:FORBIDDEN,ErrorMessage:

Through my analysis, I think it is necessary to check the code related to California jar and dtls jar to confirm why the session ID changes? please heltp to confirm it,

Thanks.

**Comment by sbernard31** [comment]
Really difficult to understand what happens. In particular when captures contain only server or device packets and when packets are truncated.  `[Packet Size limited during capture]`

My current understanding of the situation is  : 

- Client established a DTLS Connection
- Then do a register,
- Then do periodic update,
- For some reasons I didn't know, 1 update request probably timed-out. (network issue ?)
-  By default when this happens, client established a new DTLS connection trying to do a abbreviated handshake (so it use the same SessionID)
- Server could accept the abbreviated Handshake or turn it out in full-handshake (with new SessionID)
- Logs you share, make me thinks it does not accept the abbreviated Handshake (or you are using an old version of leshan client which directly do a full handshake) **==> This could be a problem in your case** but this should not be one in real world.
- Then as DTLS session is different, it seems your server refuse the update. (FORBIDDEN)
 (This is compliant with the specification v1.0.x but from my point of view this is not a good idea to link DTLS session to Registration lifetime : https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/177 . As we consider this as a bad idea [Leshan server don't that](https://github.com/eclipse/leshan/wiki/LWM2M-Observe#for-dtls.)
- As update failed, Client tries to register again (what can it do ?)

**Comment by boaks** [comment]
In my opinion, the test assumes, that, if all is OK (not too much message drops, no too much response delays, no address changes), only lwm2m "registration_updates" are to be used.

Now, for some unknown reasons, a lwm2m "registration" is used. The different session seems to be the result of a new handshake, but the root cause of that is unclear.

So, which version of the leshan lwm2m client do you use? Did you change the default configuration?

Just to mention:
I run Californium/Scandium stable for weeks, with mobile clients and DTLS CID, without any handshakes for a couple of days. So 24h should not be a general issue, but it may be hard to find the root cause, which triggers the client to do a new handshake. 
Maybe, you add some logging states to see, what triggers it.

**Comment by JameslaoA** [comment]
We use your Leshan client demo to test and find that this problem（dtlssession id is changed） will also occur. How much is the local port setting of our device more appropriate? We set the local port to 10800.

**Comment by boaks** [comment]
> We use your Leshan client demo to test and find that this problem（dtlssession id is changed） will also occur. 

That "also" makes me feel, that not only the leshan client doesn't work as expected with your server.
Anyway, if you want Simon or me to reproduce your test using a leshan client and server, we need the leshan version you use for the client.

> How much is the local port setting of our device more appropriate? We set the local port to 10800.

I don't think, this affects the test. I'm not up-to-date with leshan, but in the past there have been discussion about limiting the lifetime of a dtls session. And there are "features" as "auto-resumption", which has been miss-configured in the past.

**Comment by JameslaoA** [comment]
We use the following command to connect to the lwm2m server.
The steps are as follows:
step 1.#      java -jar leshan-client-demo-1.0.0-SNAPSHOT-jar-with-dependencies.jar -b -lp 56830 -n urn:imei:900903731526306 -i urn:imei:900903731526306 -p 6c6968793030 -u InteropBootstrap.dm.iot.att.com:5694

step 2.

**Comment by boaks** [comment]
Great!
So what I can do is to prepare a local leshan client-server setup and see, if that is able to hold the connection for 24h.

That will just help to ensure, that there is no general issue.

One remark:
Reporting issues is very welcome, but, I my opinion, it would help all, if you try to reproduce the issue with the last release.

**Comment by boaks** [comment]
I built leshan 1.0.2.
I started the server with `java -jar <server.jar>`
I added a psk device in the leshan demo server.
I started the client with `java -jar <client.jar> -u <id> -i <id> -p <hex>
Both, server and client runs on the same host.
All worked out if the box. The client sends updates every 60s.
Now see,  how long that will work ...

**Comment by JameslaoA** [comment]
> I built leshan 1.0.2.
> I started the server with `java -jar <server.jar>`
> I added a psk device in the leshan demo server.
> I started the client with `java -jar <client.jar> -u -i -p 
> Both, server and client runs on the same host.
> All worked out of the box. The client sends updates every 60s.
> Now see, how long that will work ...

okay,please feedback the test result as soon as possible.

Suggest: Our beta version(leshan-client-demo  version is V1.0.0) is recommended,the problem could be reproduced.
Thanks.

**Comment by boaks** [comment]
![Bildschirmfoto von 2020-07-10 15-03-50](https://user-images.githubusercontent.com/10937819/87157512-acfc1d80-c2be-11ea-93fe-3df210be5f83.png)

It runs now for more than 3h. All OK.
The 24h will be reached tomorrow morning, we will see, what then is shown.

**Comment by boaks** [comment]
I tried to register at AT&T, but they want a company-name for the registration.
So I have to check that with the company I work for, if I can go for that.
I executed some DTLS handshakes, and for me this looks pretty much for Californium also on their side :-).

So, my guess:
Something is not OK on the route between your leshan client and the at&t server.
Where do you execute your leshan client? Network behind a router? Does your IP-provider assign you a static IP?

**Comment by JameslaoA** [comment]
> I tried to register at AT&T, but they want a company-name for the registration.
> So I have to check that with the company I work for, if I can go for that.
> I executed some DTLS handshakes, and for me this looks pretty much for Californium also on their side :-).

You can confirm with your company that this is free registration.

> So, my guess:
> Something is not OK on the route between your leshan client and the at&t server.
> Where do you execute your leshan client? Network behind a router? Does your IP-provider assign you a static IP?

I don't think it's network or routing,i think there may be a problem with dtlssession(Californium jar), which leads to the change of session ID.

Thanks.

**Comment by JameslaoA** [comment]
> > So, my guess:
> > Something is not OK on the route between your leshan client and the at&t server.
> > Where do you execute your leshan client? Network behind a router? Does your IP-provider assign you a static IP?
>

The test Address is Florida.

**Comment by boaks** [comment]
![Bildschirmfoto von 2020-07-11 13-34-50](https://user-images.githubusercontent.com/10937819/87223209-5b659880-c37b-11ea-8acb-03940ed07356.png)

Even after a little more than 24h, in my setup, all is OK.

I'm not longer too familiar with LwM2M, anyway my conclusion:
The out of the box setup doesn't show the failure, you described, if it is ensured, that the precondition "stable static network" is fulfilled.
Why you get failures may now be caused by:
- randomly unstable client implementation 
- unstable network 

So let my point out, that a new dtls session may also be a sign of unstable network conditions.
In my experience, simply "home internet access point" causes such unstable network.
If I would run the test against the At&t from my home, my DSL-Router gets disconnected at night by my IP provider. And if the DSL router connects again, it get a new IP address. Without counter-measures, that would therefore fail to pass a 24h test.

[What Is My IP](https://www.whatismyip.com) 

So let me recommend, that you check your internet connection. Maybe just first check, if your IP is really stable over 24h.

**Comment by boaks** [comment]
> I don't think it's network or routing,i think there may be a problem with dtlssession(Californium jar), which leads to the change of session ID.

For me this looks vice versa:

In my experience the most user doesn't have a clear understand, what the network they use really looks like.
And some server use also something, which surprisingly changes the features of the network communication.
  
On the other side, Californium is the mostly used CoAP/DTLS stack for the server sides. For sure, I not only develop it now over years (as others also did, e.g. the developers of leshan also contributed to californium), I also use it now for years very successful. Communication is long-term stable, if you know, what your network does. Californium offers session resumption (also automatically or just a handshake probe, to see, if your network route has an temporary outage without losing the dtls association) and [DTLS Connection ID](https://tools.ietf.org/pdf/draft-ietf-tls-dtls-connection-id-07.pdf) to overcome ip-address changes.
I guess, even AT&T is using Eclipse Californium, at least, they support DTLS Connection ID, and there are not too many implementations out for that.

I'm not sure, which of the advanced features of Californium has made it into the LwM2M spec. If LwM2M allows to use DTLS Connection ID, and your "root cause" is a IP-change assigned by your IP-provider (see the "What Is My IP" above), then it may be as easy as adapt the line

```
DTLS_CONNECTION_ID_LENGTH=
```

in the file "Californium-properties" into
   
```
DTLS_CONNECTION_ID_LENGTH=0
```

and it may work magically. But that's just the guess, that you get more frequently a new IP than the 24h test assumes. There may be other root-causes, just investigate for them.

**Comment by boaks** [comment]
I updated the californium sandbox at californium.eclipse.org with a coap-myip-resource (see [Add MyIP](https://github.com/eclipse/californium/pull/1366) ).
It's available at the extended service using `coap://californium.eclipse.org:5783/myip` or `coaps://californium.eclipse.org:5784/myip`.

> GET coap://californium.eclipse.org:5783/myip
> ip:94.216.232.34
> port:59533

Maybe you add such coap-requests to your client. If the response changes (within 24h), your network isn't stable/static enough to execute a 24h test.

**Comment by boaks** [comment]
@JameslaoA 

Any results from verifying your "long term ip-address"?
Either with [HTTP - What Is my IP](https://www.whatismyip.com/) or with `coap://californium.eclipse.org:5783/myip` ?

**Comment by JameslaoA** [comment]
Can you test it with v1.0.0?  The V1.0.0 seems to have this problem.

**Comment by boaks** [comment]
Does this mean, a 1.0.2 solves this issue?
Meaning, if you test with 1.0.2 you pass the test?

**Comment by JameslaoA** [comment]
> Does this mean, a 1.0.2 solves this issue?
> Meaning, if you test with 1.0.2 you pass the test?

Yes, V1.0.2 doesn't seem to have this problem. Let's check the version of Californium from v2.1.0 to v2.1.3.

**Comment by boaks** [comment]
I don't think, that I have enough time to spend into analyze things, which seems to be already solved :-).

**Comment by JameslaoA** [comment]
I think it is necessary to find the root cause of the problem, which will help us to solve the problem positively.

**Comment by boaks** [comment]
![Bildschirmfoto von 2020-07-15 10-36-33](https://user-images.githubusercontent.com/10937819/87523820-b3f0aa80-c687-11ea-9929-92b6d847ebce.png)

Retested with 1.0.0 - works 24h stable as well. From my side, it's done.

**Comment by boaks** [comment]
Just to mention:
To register at the AT&T service requires the "legal responsible agreement" of your company. 
Completely overdone! 
And without complete that data nightmare, you can't even remove the "half registration".
So, I really out of this issue.

**Comment by weicanqd** [comment]
hi， i had try  test the  RegisterUpdate for many times, i add  the log and catched in californium  scandium 2.1.0 as below:

07-16 06:48:17.628  2354  3317 D lihy lwm2m: RequestSender
07-16 06:48:17.634  2354  3317 D lihy lwm2m: CoapEndpoint request.str:CON-POST   MID=   -1, Token=null, OptionSet={"Uri-Path":["rd","urn:imei:900903731526306"]}, no payload
07-16 06:48:17.636  2354  3317 D lihy lwm2m: CaliforniumLwM2mRequestSender RequestSender end
07-16 06:48:17.640  2354  3316 D lihy lwm2m: send(final RawData message) 01
07-16 06:48:17.640  2354  3316 D lihy lwm2m: send(final RawData message) 02 getConnection
07-16 06:48:17.640  2354  3316 D lihy lwm2m: getConnection cid null
07-16 06:48:17.640  2354  3316 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:17.640  2354  3316 D lihy lwm2m: InMemoryConnectionStore get()  missing connection for peerAddress
07-16 06:48:17.641  2354  3316 D lihy lwm2m: getConnection create new connection for peerAddress create :true
07-16 06:48:17.641  2354  3316 D lihy lwm2m: Connection 2 parm
07-16 06:48:17.641  2354  3316 D lihy lwm2m: InMemoryConnectionStore put connection: add {}
07-16 06:48:17.642  2354  3316 D lihy lwm2m: BaseCoapStack sendRequest
07-16 06:48:17.643  2354  3362 D lihy lwm2m: send(final RawData message) 05 sendMessage
07-16 06:48:17.643  2354  3362 D lihy lwm2m: sendMessage Sending application layer message
07-16 06:48:17.643  2354  3362 D lihy lwm2m: sendMessage handshaker == null
07-16 06:48:17.645  2354  3362 D lihy lwm2m: client_hello sessionId :null
07-16 06:48:17.898  2354  3363 D lihy lwm2m: getConnection cid null
07-16 06:48:17.898  2354  3363 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:17.900  2354  3362 D lihy lwm2m: processDatagram 04
07-16 06:48:17.901  2354  3362 D lihy lwm2m: processRecord HANDSHAKE
07-16 06:48:18.178  2354  3363 D lihy lwm2m: getConnection cid null
07-16 06:48:18.178  2354  3363 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:18.179  2354  3362 D lihy lwm2m: processDatagram 04
07-16 06:48:18.182  2354  3362 D lihy lwm2m: ServerHello random :                       GMT Unix Time: Thu Jul 16 06:48:18 GMT 2020
07-16 06:48:18.182  2354  3362 D lihy lwm2m:                    Random Bytes: A0B94F8EC9A71416671090C963A8FEBEA84BD705EC5B7B681CCAD54F
07-16 06:48:18.182  2354  3362 D lihy lwm2m: sessionid :5F0FF8324C7A997A11FE7A4284F40C1D48C5AC02E4C751D1C4D25A1D1B132436
07-16 06:48:18.183  2354  3362 D lihy lwm2m: processRecord HANDSHAKE
07-16 06:48:18.184  2354  3362 D lihy lwm2m: processDatagram 04
07-16 06:48:18.185  2354  3362 D lihy lwm2m: processRecord HANDSHAKE
07-16 06:48:19.204  2354  3362 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:19.496  2354  3363 D lihy lwm2m: getConnection cid null
07-16 06:48:19.496  2354  3363 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:19.498  2354  3362 D lihy lwm2m: processDatagram 04
07-16 06:48:19.498  2354  3362 D lihy lwm2m: processRecord CHANGE_CIPHER_SPEC
07-16 06:48:19.499  2354  3362 D lihy lwm2m: processDatagram 04
07-16 06:48:19.503  2354  3362 D lihy lwm2m: processRecord HANDSHAKE
07-16 06:48:19.780  2354  3363 D lihy lwm2m: getConnection cid null
07-16 06:48:19.780  2354  3363 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:19.782  2354  3362 D lihy lwm2m: processDatagram 04
07-16 06:48:19.785  2354  3362 D lihy lwm2m: processRecord APPLICATION_DATA
07-16 06:48:19.785  2354  3362 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:19.792  2354  3317 D lihy lwm2m: Registration update failed: code:FORBIDDEN,ErrorMessage:

so we can see the connection lost(missed by get by peerAddress) when send the Registration update request, so the client would be create a new connection whth a new session id , but the server check the session id  that is different with the last Registration update connection session id . so the server will send 
code:FORBIDDEN for response

i wan to to know  why the connection would been lost , the peeraddrss change or  other reseon??

**Comment by boaks** [comment]
> we can see the connection lost(missed by get by peerAddress)

I'm not sure, which line(s) in your log should show that?

Generally there are two ways to see, what causes the issue: 
either add  on the client side a GET requests to coap://californium.eclipse.org:5783/myip and check, if the result is changing,
or add a check on the server side, if a "new registration" also comes with a new address and fail the 24h test with the hint, that it may have failed because of the address change.

if you use Californium (or an other DTLS cid implementation), this may already work even with address changes if you enable to use CID (see above [DTLS_CONNECTION_ID_LENGTH=0](https://github.com/eclipse/leshan/issues/863#issuecomment-657056859) ).
But let me say, if there is a disruption for 1-2 minutes (what may be usual, if your home-ip-provider disconnects you at night), this may cause additional trouble, where dtls cid alone doesn't help. 
An other way to check would be to run the lwm2m client from a cloud v-box, that should have a static ip-connection.

**Comment by boaks** [comment]
@sbernard31 
I'm not longer too common with LwM2M.
Is a "update registration" "FORBIDDEN", if it comes with a different address?
Or only, if it comes with a different DTLS Session?

**Comment by weicanqd** [comment]
thankyou for your analysisn.

client will get the current connect when send a new requestion, the log info get connect as below:
07-16 06:48:17.640 2354 3316 D lihy lwm2m: InMemoryConnectionStore get() connection by peerAddress
07-16 06:48:17.640 2354 3316 D lihy lwm2m: InMemoryConnectionStore get() missing connection for peerAddress
but the connection missed,so it will create a new connection and create a new session id.

now, i'm sure my client ip is not change ,the server ip is not change  and the newwork is ok ,
so i  can't find the reseon the connection lost.

i will go on.

**Comment by boaks** [comment]
Strange, that log is from your client? Or server?

**Comment by boaks** [comment]
> nslookup InteropBootstrap.dm.iot.att.com
> 
> Non-authoritative answer:
> InteropBootstrap.dm.iot.att.com	canonical name = interopbootstrap.iot.attcompute.com.
> Name:	interopbootstrap.iot.attcompute.com
> Address: 13.91.120.235
> Name:	interopbootstrap.iot.attcompute.com
> Address: 13.91.132.98
> Name:	interopbootstrap.iot.attcompute.com
> Address: 13.91.132.104
> 

:-) 

I guess, it is important, that the lwm2m client does only on DNS lookup and then use that IP-address for 24h ...
Or you adapt your host to use a static local dns setup, which only contains one of the addresses.

**Comment by sbernard31** [comment]
After reading all this thread again, I still bet on my previous explanation : https://github.com/eclipse/leshan/issues/863#issuecomment-656119753

> i wan to to know why the connection would been lost

I try to explain this a bit more :  
The UPDATE probably timed-out after several CoAP retries (Default time-out 2min).
Why the request timed-out ?  many possible reasons : 
  1. Server lost its connection for any reason (connection lifetime expires, server reboot ...)
  2. Client IP changed and so server will not be able to retrieve the corresponding connection.
  3. Network issues. 

  For 1. and 2. the solution is to create new DTLS connection. This is why Leshan client will try to send again the UPDATE with a new connection (by removing the previous one and so trigger a new connection creation trying to reuse the same session).
  For 3. create a new connection will not help. 

  For 1. there is no other solution than create a new connection.
  For 2. as @boaks say connectionID could help.
  For 3. 2minutes timeout + CoAP retries should limit to create connection for bad reason. If this is not enough you can change all the [timeout](https://github.com/eclipse/leshan/wiki/Request-Timeout) (LWM2M or CoAP one). Maybe using PROBE mode could help but in this case it seems to me that this will not :thinking:.


> Is a "update registration" "FORBIDDEN", if it comes with a different address?

As I said , I guess FORBIDDEN is because SESSION ID changed. The LWM2M specification 1.0.x says : _"When a new DTLS Session is started, or in NoSec mode when the LwM2MClient IP address changes, the Client MUST register again to the LwM2MServer"_  (Currently, I consider this is a bad idea : https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/177#issuecomment-267590843)

When Leshan try to create a new connection it SHOULD try to resume previous session. So another question is : 
- Do you see that Leshan use the correct SessionID  ? (As @boacks, I'm not really interested in debugging an old version of Leshan so please use the 1.0.2)
- Is server accept this SessionID ?

**Comment by sbernard31** [comment]
If all is triggered by an UPDATE request timeout, you should see this [log](https://github.com/eclipse/leshan/blob/leshan-1.0.2/leshan-client-core/src/main/java/org/eclipse/leshan/client/engine/DefaultRegistrationEngine.java#L422).

**Comment by boaks** [comment]
@JameslaoA  and @weicanqd 

If you both (or at least one) testing against AT&T, please check, what you get on bootstrap.
If it's "InteropBootstrap.dm.iot.att.com",  the issue is just, that you may get different destinations when using it.
if it's one of the "literal IPs", then this is not the root cause.

**Comment by boaks** [comment]
[Java DNS Cache TTL](https://docs.oracle.com/javase/7/docs/technotes/guides/net/properties.html)

Maybe trying to adjust the `networkaddress.cache.ttl` accordingly (24h := 100000, or -1 ) helps. But I never tested it on my own.

**Comment by sbernard31** [comment]
>  If it's "InteropBootstrap.dm.iot.att.com", the issue is just, that you may get different destinations when using it.

If server IP changes, this is probably visible in a tcpdump capture at client side.

**Comment by boaks** [comment]
Seems to be an "issue nest".
May be changing client addresses,
May be changing server addresses,
May be communication outages.

**Comment by boaks** [comment]
**Communication is long-term stable, if you know, what your network does.**

From @JameslaoA capture:

![Bildschirmfoto von 2020-07-18 08-15-54](https://user-images.githubusercontent.com/10937819/87846280-125b9a00-c8cf-11ea-987a-50434baba443.png)

> nslookup InteropLwM2M.dm.iot.att.com
> Server:		127.0.0.53
> Address:	127.0.0.53#53
> 
> Non-authoritative answer:
> InteropLwM2M.dm.iot.att.com	canonical name = interoplwm2m.iot.attcompute.com.
> Name:	interoplwm2m.iot.attcompute.com
> Address: 13.91.132.104
> Name:	interoplwm2m.iot.attcompute.com
> Address: 13.91.120.235
> Name:	interoplwm2m.iot.attcompute.com
> Address: 13.91.132.98
> 

The capture shows, that after the DNS request, the client assumes an other server ip-address.

**Comment by weicanqd** [comment]
yes ,i make sure the server side has three ip address for the same url.
       i check the register update code  and find that will get the ip address everytime by "server.getIdentity().getPeerAddress()", so we could been got an other ip-address. 

the register update code ;
DeregisterResponse response = sender.send(server.getIdentity().getPeerAddress(),
                    server.getIdentity().isSecure(), new DeregisterRequest(registrationID), deregistrationTimeoutInMs);

so we get the root reseon for my question.
thankyou verymuch!

**Comment by weicanqd** [comment]
the update requestion code, get ip by dmInfo.getAddress() every time.

UpdateResponse response = sender.send(dmInfo.getAddress(), dmInfo.isSecure(),
                    new UpdateRequest(registrationID, registrationUpdate.getLifeTimeInSec(),
                            registrationUpdate.getSmsNumber(), registrationUpdate.getBindingMode(),
                            registrationUpdate.getObjectLinks(), registrationUpdate.getAdditionalAttributes()),
                    requestTimeoutInMs);

**Comment by boaks** [comment]
You're welcome.

Just to check, if there is more unexpected ongoing in the network, please try to adjust the [Java DNS Cache TTL](https://docs.oracle.com/javase/7/docs/technotes/guides/net/properties.html) as proposed in [Comment above](https://github.com/eclipse/leshan/issues/863#issuecomment-659281405) . 
If that works, that ensures, that you're not faced additional issues.

**Comment by sbernard31** [comment]
Good to finally know what happened.

```java
UpdateResponse response = sender.send(dmInfo.getAddress(), dmInfo.isSecure(),
new UpdateRequest(registrationID, registrationUpdate.getLifeTimeInSec(),
registrationUpdate.getSmsNumber(), registrationUpdate.getBindingMode(),
registrationUpdate.getObjectLinks(), registrationUpdate.getAdditionalAttributes()),
requestTimeoutInMs);
```
This code changed a lot since the 1.0.0-SNAPSHOT version you download  the _2020-3-1_. 
For the future I really advice you to use recent patched version (1.0.x or 1.1.x or ...) instead of an old and random master one.
(You will probably face less issue or you will help to detect and fix new issue rather than old one)

Reading the code and your comments above (https://github.com/eclipse/leshan/issues/863#issuecomment-658003963), I understand this works as expected for leshan 1.0.2 now. Could you confirm ?
Can we close this issue now ?

**Comment by weicanqd** [comment]
yes, this issue can been closed , thankyou !
